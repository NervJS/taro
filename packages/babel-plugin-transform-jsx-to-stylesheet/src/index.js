const path = require('path')

const STYLE_SHEET_NAME = '_styleSheet'
const GET_STYLE_FUNC_NAME = '_getStyle'
const MERGE_STYLES_FUNC_NAME = '_mergeStyles'
const GET_CLS_NAME_FUNC_NAME = '_getClassName'
const NAME_SUFFIX = 'StyleSheet'
const cssSuffixs = ['.css', '.scss', '.sass', '.less', '.styl']
let styleNames = [] // css module 写法中 import 的 Identifier

module.exports = function ({types: t, template}) {
  const mergeStylesFunctionTemplate = template(`
function ${MERGE_STYLES_FUNC_NAME}() {
  var newTarget = {};

  for (var index = 0; index < arguments.length; index++) {
    var target = arguments[index];

    for (var key in target) {
      newTarget[key] = Object.assign(newTarget[key] || {}, target[key]);
    }
  }

  return newTarget;
}
  `)
  const getClassNameFunctionTemplate = template(`
function ${GET_CLS_NAME_FUNC_NAME}() {
  var className = [];
  var args = arguments[0];
  var type = Object.prototype.toString.call(args).slice(8, -1).toLowerCase();

  if (type === 'string') {
    args = args.trim();
    args && className.push(args);
  } else if (type === 'array') {
    args.forEach(function (cls) {
      cls = ${GET_CLS_NAME_FUNC_NAME}(cls).trim();
      cls && className.push(cls);
    });
  } else if (type === 'object') {
    for (var k in args) {
      k = k.trim();
      if (k && args.hasOwnProperty(k) && args[k]) {
        className.push(k);
      }
    }
  }

  return className.join(' ').trim();
}
  `)
  const getStyleFunctionTemplete = template(`
function ${GET_STYLE_FUNC_NAME}(classNameExpression) { 
  var className = ${GET_CLS_NAME_FUNC_NAME}(classNameExpression);
  var classNameArr = className.split(/\\s+/);

  var style = [];
  if (classNameArr.length === 1) {
    style.push(${STYLE_SHEET_NAME}[classNameArr[0].trim()]);
  } else {
      classNameArr.forEach(function(cls) {
      style.push(${STYLE_SHEET_NAME}[cls.trim()]);
    });
  }

  return style;
}
  `)

  const getClassNameFunctionAst = getClassNameFunctionTemplate()
  const mergeStylesFunctionAst = mergeStylesFunctionTemplate()
  const getStyleFunctionAst = getStyleFunctionTemplete()

  function getArrayExpression (value) {
    let expression
    let str

    if (!value || value.value === '') {
      // className
      // className=""
      return []
    } else if (value.type === 'JSXExpressionContainer' && value.expression && typeof value.expression.value !== 'string') {
      // className={{ container: true }}
      // className={['container wrapper', { scroll: false }]}
      return [t.callExpression(t.identifier(GET_STYLE_FUNC_NAME), [value.expression])]
    } else {
      // className="container"
      // className={'container'}
      str = (value.expression ? value.expression.value : value.value).trim()
    }

    return str === '' ? [] : str.split(/\s+/).map((className) => {
      return template(`${STYLE_SHEET_NAME}["${className}"]`)().expression
    })
  }

  function findLastImportIndex (body) {
    const bodyReverse = body.slice(0).reverse()
    let _index = 0

    bodyReverse.some((node, index) => {
      if (node.type === 'ImportDeclaration') {
        _index = body.length - index - 1
        return true
      }
      return false
    })

    return _index
  }

  return {
    visitor: {
      Program: {
        exit (astPath, {file}) {
          const node = astPath.node
          // const cssFileCount = file.get('cssFileCount')
          const injectGetStyle = file.get('injectGetStyle')
          const lastImportIndex = findLastImportIndex(node.body)
          let cssParamIdentifiers = file.get('cssParamIdentifiers')
          let callExpression

          if (cssParamIdentifiers) {
            // only one css file，由于样式文件合并，永远只有一个
            if (cssParamIdentifiers.length === 1) {
              callExpression = t.variableDeclaration('var', [t.variableDeclarator(t.identifier(STYLE_SHEET_NAME), cssParamIdentifiers[0])])
            } else if (cssParamIdentifiers.length > 1) {
              const objectAssignExpression = t.callExpression(t.identifier(MERGE_STYLES_FUNC_NAME), cssParamIdentifiers)
              callExpression = t.variableDeclaration('var', [t.variableDeclarator(t.identifier(STYLE_SHEET_NAME), objectAssignExpression)])
            }

            node.body.splice(lastImportIndex + 1, 0, callExpression)

            if (injectGetStyle) {
              node.body.splice(lastImportIndex + 2, 0, getClassNameFunctionAst)
              node.body.splice(lastImportIndex + 3, 0, getStyleFunctionAst)
            }
          }

          // 全部合并成一个，也就不需要 mergeStylesFunctionAst 了
          // if (cssFileCount > 1) {
          //   node.body.unshift(mergeStylesFunctionAst)
          // }

          // 将 css modules 的 styles 替换为 _styleSheet
          astPath.traverse({
            JSXOpeningElement (astPath) {
              astPath.traverse({
                JSXAttribute ({node}) {
                  if (node.name && node.name.name === 'style') {
                    astPath.traverse({
                      MemberExpression ({node}) {
                        if (node.object.type === 'Identifier' && styleNames.indexOf(node.object.name) > -1) {
                          node.object.name = STYLE_SHEET_NAME
                        }
                      }
                    })
                  }
                }
              })
            }
          })
          // 清除该 js 文件的 styleNames
          styleNames = []
        }
      },
      JSXOpeningElement ({container}, {file}) {
        const cssFileCount = file.get('cssFileCount') || 0
        if (cssFileCount < 1) {
          return
        }

        // Check if has "style"
        let hasStyleAttribute = false
        let styleAttribute
        let hasClassName = false
        let classNameAttribute

        const attributes = container.openingElement.attributes
        for (let i = 0; i < attributes.length; i++) {
          const name = attributes[i].name
          if (name) {
            if (!hasStyleAttribute) {
              hasStyleAttribute = name.name === 'style'
              styleAttribute = hasStyleAttribute && attributes[i]
            }

            if (!hasClassName) {
              hasClassName = name.name === 'className'
              classNameAttribute = hasClassName && attributes[i]
            }
          }
        }

        if (hasClassName) {
          // Remove origin className
          attributes.splice(attributes.indexOf(classNameAttribute), 1)

          if (
            classNameAttribute.value &&
            classNameAttribute.value.type === 'JSXExpressionContainer' &&
            typeof classNameAttribute.value.expression.value !== 'string' // not like className={'container'}
          ) {
            file.set('injectGetStyle', true)
          }

          const arrayExpression = getArrayExpression(classNameAttribute.value)

          if (arrayExpression.length === 0) {
            return
          }

          if (hasStyleAttribute && styleAttribute.value) {
            let expression = styleAttribute.value.expression
            let expressionType = expression.type

            // style={[styles.a, styles.b]} ArrayExpression
            if (expressionType === 'ArrayExpression') {
              expression.elements = arrayExpression.concat(expression.elements)
              // style={styles.a} MemberExpression
              // style={{ height: 100 }} ObjectExpression
              // style={{ ...custom }} ObjectExpression
              // style={custom} Identifier
              // style={getStyle()} CallExpression
              // style={this.props.useCustom ? custom : null} ConditionalExpression
              // style={custom || other} LogicalExpression
            } else {
              styleAttribute.value.expression = t.arrayExpression(arrayExpression.concat(expression))
            }
          } else {
            let expression = arrayExpression.length === 1 ? arrayExpression[0] : t.arrayExpression(arrayExpression)
            attributes.push(t.jSXAttribute(t.jSXIdentifier('style'), t.jSXExpressionContainer(expression)))
          }
        }
      },
      // 由于目前 js 引入的文件样式默认会全部合并，故进插入一个就好，其余的全部 remove
      ImportDeclaration (astPath, state) {
        const {file} = state
        const node = astPath.node
        const sourceValue = node.source.value
        const specifiers = node.specifiers
        const jsFilePath = state.opts.filePath // 传进来的文件的 filePath ,Babel6 file.opts.filaname 为unknown
        const extname = path.extname(sourceValue)
        const cssIndex = cssSuffixs.indexOf(extname)
        let cssFileCount = file.get('cssFileCount') || 0
        let cssParamIdentifiers = file.get('cssParamIdentifiers') || []

        if (cssIndex > -1) {
          // `import styles from './foo.css'` kind
          if (node.importKind === 'value' && specifiers.length > 0) {
            styleNames.push(specifiers[0].local.name)
          }
          // 第一个引入的样式文件
          if (cssFileCount === 0) {
            const cssFileBaseName = path.basename(jsFilePath, path.extname(jsFilePath))
            // 引入样式对应的变量名
            const styleSheetIdentifierValue = `${cssFileBaseName.replace(/-/g, '_') + NAME_SUFFIX}`
            const styleSheetIdentifierPath = `./${cssFileBaseName}_styles`
            const styleSheetIdentifier = t.identifier(styleSheetIdentifierValue)

            node.specifiers = [t.importDefaultSpecifier(styleSheetIdentifier)]
            node.source = t.stringLiteral(styleSheetIdentifierPath)
            cssParamIdentifiers.push(styleSheetIdentifier)
          } else {
            astPath.remove()
          }
          cssFileCount++

          file.set('cssParamIdentifiers', cssParamIdentifiers)
          file.set('cssFileCount', cssFileCount)
        }
      }
    }
  }
}
