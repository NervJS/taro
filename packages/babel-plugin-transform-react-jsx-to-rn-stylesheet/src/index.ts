import path from 'path'
import camelize from 'camelize'
import { transformCSS } from 'taro-css-to-react-native'
import { types as Types, template as Template, PluginObj } from 'babel__core'
import { ConvertPluginPass as PluginPass } from './types'

const STYLE_SHEET_NAME = '_styleSheet'
const GET_STYLE_FUNC_NAME = '_getStyle'
const MERGE_STYLES_FUNC_NAME = '_mergeStyles'
const GET_CLS_NAME_FUNC_NAME = '_getClassName'
const NAME_SUFFIX = 'styleSheet'
const RN_CSS_EXT = ['.css', '.scss', '.sass', '.less', '.styl', '.stylus']

const isStyle = value => {
  const ext = path.extname(value)
  return RN_CSS_EXT.indexOf(ext) > -1
}

const isModuleSource = value => value.indexOf('.module.') > -1

const string2Object = str => {
  const entries = str.replace(/;+$/g, '')
    .split(';')
    .map(l => {
      const arr = l.split(':')
      arr[1] = arr[1]?.replace(/px/g, 'PX')
      return arr
    })
  const cssObject = transformCSS(entries)
  return cssObject
}

const object2Expression = (template, cssObject) => {
  const ast = template.ast(`var a = ${JSON.stringify(cssObject)}`)
  return ast.declarations[0]?.init
}

function findLastImportIndex (body) {
  const bodyReverse = body.slice(0).reverse()
  let _index = 0

  bodyReverse.some((node, index) => {
    if (node.type === 'ImportDeclaration') {
      _index = body.length - index - 1
      return true
    }
    return false
  })

  return _index
}

const MergeStylesFunction = `
function _mergeStyles() {
  var newTarget = {};

  for (var index = 0; index < arguments.length; index++) {
    var target = arguments[index];

    for (var key in target) {
      newTarget[key] = Object.assign(newTarget[key] || {}, target[key]);
    }
  }

  return newTarget;
}
`
const getClassNameFunction = `
function ${GET_CLS_NAME_FUNC_NAME}() {
  var className = [];
  var args = arguments[0];
  var type = Object.prototype.toString.call(args).slice(8, -1).toLowerCase();

  if (type === 'string') {
    args = args.trim();
    args && className.push(args);
  } else if (type === 'array') {
    args.forEach(function (cls) {
      cls = ${GET_CLS_NAME_FUNC_NAME}(cls).trim();
      cls && className.push(cls);
    });
  } else if (type === 'object') {
    for (var k in args) {
      k = k.trim();
      if (k && args.hasOwnProperty(k) && args[k]) {
        className.push(k);
      }
    }
  }

  return className.join(' ').trim();
}
`
const getStyleFunction = `
function ${GET_STYLE_FUNC_NAME}(classNameExpression) { 
  var className = ${GET_CLS_NAME_FUNC_NAME}(classNameExpression);
  var classNameArr = className.split(/\\s+/);

  var style = [];
  if (classNameArr.length === 1) {
    style.push(${STYLE_SHEET_NAME}[classNameArr[0].trim()]);
  } else {
      classNameArr.forEach(function(cls) {
      style.push(${STYLE_SHEET_NAME}[cls.trim()]);
    });
  }

  return style;
}
`

export default function (babel: {
  types: typeof Types
  template: typeof Template
}): PluginObj {
  const { types: t, template } = babel

  const getClassNameFunctionTemplate = template(getClassNameFunction)
  const getStyleFunctionTemplete = template(getStyleFunction)
  const getClassNameFunctionStmt = getClassNameFunctionTemplate()
  const getStyleFunctionStmt = getStyleFunctionTemplete()

  function getMap (str) {
    return str.split(/\s+/).map((className) => {
      // return template(`${STYLE_SHEET_NAME}["${className}"]`)().expression
      const stmt = template(`${STYLE_SHEET_NAME}["${className}"]`)()
      if (t.isExpressionStatement(stmt)) {
        return stmt.expression
      }
    })
  }

  function isCSSModuleExpression (value, cssModuleStylesheets) {
    if (t.isJSXExpressionContainer(value)) {
      if (t.isMemberExpression(value.expression) && t.isIdentifier(value.expression.object)) {
        // className 表达式包括了 css module 引用时处理成 style 属性
        if (cssModuleStylesheets.includes(value.expression.object.name)) {
          return true
        }
      }
    }
  }

  function getArrayExpression (value, cssModuleStylesheets) {
    // css module 时 className 处理成 style 属性，所以直接取值跟 style 合并
    if (isCSSModuleExpression(value, cssModuleStylesheets)) {
      return [value.expression]
    }

    let str

    if (!value || value.value === '') {
      // className
      // className=""
      return []
    } else if (value.type === 'JSXExpressionContainer' && value.expression && typeof value.expression.value !== 'string') {
      // className={{ container: true }}
      // className={['container wrapper', { scroll: false }]}
      return [t.callExpression(t.identifier(GET_STYLE_FUNC_NAME), [value.expression])]
    } else {
      // className="container"
      // className={'container'}
      str = (value.expression ? value.expression.value : value.value).trim()
    }

    return str === '' ? [] : getMap(str)
  }

  let existStyleImport = false

  return {
    name: 'transform-react-jsx-to-rn-stylesheet',
    visitor: {
      Program: {
        enter (astPath, state: PluginPass) {
          let lastImportAstPath
          for (const stmt of astPath.get('body')) {
            if (t.isImportDeclaration(stmt.node)) {
              if (isStyle(stmt.node.source.value)) {
                existStyleImport = true
                // 在 Program 最开始处理 import 样式
                // 防止有些插件在入口对 import 进行处理，比如 @babel/plugin-transform-typescript 会进行删除
                importDeclaration(stmt, state, t)
              }
              lastImportAstPath = stmt
            }
          }
          if (existStyleImport) {
            const { file } = state
            const styleSheetIdentifiers = file.get('styleSheetIdentifiers') || []
            let expression
            // only one css file，由于样式文件合并，永远只有一个
            if (styleSheetIdentifiers.length === 1) {
              expression = `var ${STYLE_SHEET_NAME} = ${styleSheetIdentifiers[0].name};\n`
            } else if (styleSheetIdentifiers.length > 1) {
              const params = styleSheetIdentifiers.reduce((current, next) => `${current},${next.name}`, '').slice(1)
              expression = `${MergeStylesFunction}\n
              var ${STYLE_SHEET_NAME} = ${MERGE_STYLES_FUNC_NAME}(${params});\n`
            } else {
              expression = `var ${STYLE_SHEET_NAME} = {}`
            }
            const expressionAst = template.ast(expression)
            lastImportAstPath.insertAfter(expressionAst)
          }
        },
        exit (astPath, state: PluginPass) {
          const { file } = state
          const node = astPath.node
          const injectGetStyle = file.get('injectGetStyle')
          // 从最后一个import 开始插入表达式，后续插入的表达式追加在后面
          const lastImportIndex = findLastImportIndex(node.body)

          if (injectGetStyle) {
            // @ts-ignore
            node.body.splice(lastImportIndex + 1, 0, getClassNameFunctionStmt)
            // @ts-ignore
            node.body.splice(lastImportIndex + 2, 0, getStyleFunctionStmt)
          }
          existStyleImport = false
        }
      },
      JSXOpeningElement ({ node }, state: PluginPass) {
        const { file } = state
        const cssModuleStylesheets = file.get('cssModuleStylesheets') || []

        // Check if has "style"
        let hasStyleAttribute = false
        let styleAttribute
        let hasClassName = false
        let classNameAttribute

        const attributes = node.attributes
        for (let i = 0; i < attributes.length; i++) {
          const attribute = attributes[i]
          if (t.isJSXAttribute(attribute)) {
            const name = attribute.name
            if (name) {
              if (!hasStyleAttribute) {
                hasStyleAttribute = name.name === 'style'
                styleAttribute = hasStyleAttribute && attributes[i]
              }

              if (!hasClassName) {
                hasClassName = name.name === 'className'
                classNameAttribute = hasClassName && attributes[i]
              }
            }
          }
        }

        if (hasClassName && existStyleImport) {
          // Remove origin className
          attributes.splice(attributes.indexOf(classNameAttribute), 1)

          if (
            classNameAttribute.value &&
            classNameAttribute.value.type === 'JSXExpressionContainer' &&
            typeof classNameAttribute.value.expression.value !== 'string' && // not like className={'container'}
            !isCSSModuleExpression(classNameAttribute.value, cssModuleStylesheets) // 不含有 css module 变量的表达式
          ) {
            file.set('injectGetStyle', true)
          }

          const arrayExpression = getArrayExpression(classNameAttribute.value, cssModuleStylesheets)

          if (arrayExpression.length === 0) {
            return
          }

          if (hasStyleAttribute && styleAttribute.value) {
            let expression
            // 支持 行内 style 转成oject：style="width:100;height:100;" => style={{width:'100',height:'100'}}
            if (t.isStringLiteral(styleAttribute.value)) {
              const cssObject = string2Object(styleAttribute.value.value)
              expression = object2Expression(template, cssObject)
            } else {
              expression = styleAttribute.value.expression
            }
            const expressionType = expression.type

            // style={[styles.a, styles.b]} ArrayExpression
            if (expressionType === 'ArrayExpression') {
              expression.elements = arrayExpression.concat(expression.elements)
              // style={styles.a} MemberExpression
              // style={{ height: 100 }} ObjectExpression
              // style={{ ...custom }} ObjectExpression
              // style={custom} Identifier
              // style={getStyle()} CallExpression
              // style={this.props.useCustom ? custom : null} ConditionalExpression
              // style={custom || other} LogicalExpression
            } else {
              styleAttribute.value = t.jSXExpressionContainer(t.arrayExpression(arrayExpression.concat(expression)))
            }
          } else {
            const expression = arrayExpression.length === 1 ? arrayExpression[0] : t.arrayExpression(arrayExpression)
            attributes.push(t.jSXAttribute(t.jSXIdentifier('style'), t.jSXExpressionContainer(expression)))
          }
        } else if (hasStyleAttribute) {
          if (t.isStringLiteral(styleAttribute.value)) {
            const cssObject = string2Object(styleAttribute.value.value)
            styleAttribute.value = t.jSXExpressionContainer(object2Expression(template, cssObject))
          }
        }
      }
    }
  }
}

// 由于目前 js 引入的文件样式默认会全部合并，故进插入一个就好，其余的全部 remove
function importDeclaration (astPath, state, t) {
  const { file } = state
  const node = astPath.node
  const sourceValue = node.source.value
  const specifiers = node.specifiers
  const ext = path.extname(sourceValue)
  const styleSheetIdentifiers = file.get('styleSheetIdentifiers') || []
  const cssModuleStylesheets = file.get('cssModuleStylesheets') || []
  const isCSSModule = state.opts?.isCSSModule

  // 是样式文件但不是 css module
  if (isStyle(sourceValue)) {
    // 导入的样式赋值
    let styleSheetName = ''
    // `import styles from './foo.css'` kind
    if (specifiers.length > 0) {
      styleSheetName = specifiers[0].local.name
    }

    if (isModuleSource(sourceValue) && isCSSModule) {
      if (styleSheetName) {
        cssModuleStylesheets.push(styleSheetName)
      }
    } else {
      const cssFileName = path.basename(sourceValue)
      const cssFileBaseName = path.basename(cssFileName, ext)
      // 引入样式对应的变量名
      const styleSheetSource = sourceValue
      let styleSheetIdentifierName
      if (styleSheetName) {
        styleSheetIdentifierName = styleSheetName
      } else {
        styleSheetName = camelize(`${cssFileBaseName}${ext}_${NAME_SUFFIX}`)
        const repeatName = styleSheetIdentifiers.find(identifier => identifier.name === styleSheetName)
        styleSheetIdentifierName = repeatName ? styleSheetName + '1' : styleSheetName // fix repeat name
        // styleSheetIdentifierName = camelize(`${cssFileBaseName}${ext}_${NAME_SUFFIX}`)
      }
      const styleSheetIdentifier = t.identifier(styleSheetIdentifierName)

      node.specifiers = [t.importDefaultSpecifier(styleSheetIdentifier)]
      node.source = t.stringLiteral(styleSheetSource)
      styleSheetIdentifiers.push(styleSheetIdentifier)
    }
  }

  file.set('styleSheetIdentifiers', styleSheetIdentifiers)
  file.set('cssModuleStylesheets', cssModuleStylesheets)
}
