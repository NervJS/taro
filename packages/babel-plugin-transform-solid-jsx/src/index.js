import SyntaxJSX from '@babel/plugin-syntax-jsx'

import postprocess from './shared/postprocess'
import preprocess from './shared/preprocess'
import { transformJSX } from './shared/transform'

export default () => {
  return {
    name: 'JSX DOM Expressions',
    inherits: SyntaxJSX.default,
    visitor: {
      JSXElement: transformJSX,
      JSXFragment: transformJSX,
      Program: {
        enter: preprocess,
        exit: postprocess
      }
    }
  }
}
