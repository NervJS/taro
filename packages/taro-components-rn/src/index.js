export { default as View } from './components/View'
export { default as ScrollView } from './components/ScrollView'
export { default as Swiper } from './components/Swiper'
export { default as SwiperItem } from './components/Swiper/SwiperItem'
export { default as Icon } from './components/Icon'
export { default as Text } from './components/Text'
export { default as RichText } from './components/RichText'
export { default as Progress } from './components/Progress'
export { default as Button } from './components/Button'
export { default as Checkbox } from './components/Checkbox'
export { default as CheckboxGroup } from './components/Checkbox/checkbox-group'
export { default as Form } from './components/Form'
export { default as Input } from './components/Input'
export { default as Label } from './components/Label'
export { default as Picker } from './components/Picker'
export { default as Radio } from './components/Radio'
export { default as RadioGroup } from './components/Radio/radio-group'
export { default as Slider } from './components/Slider'
export { default as Switch } from './components/Switch'
export { default as Textarea } from './components/Textarea'
export { default as Image } from './components/Image'
export { default as Tabbar } from './components/Tabbar'
export { default as TabbarContainer } from './components/Tabbar/Container'
export { default as TabbarPanel } from './components/Tabbar/Panel'
export { default as Audio } from './components/Audio'
export { default as Video } from './components/Video'
export { default as Map } from './components/Map'
