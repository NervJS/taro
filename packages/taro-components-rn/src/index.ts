export { default as Provider } from './lib/provider'

// export { default as Audio } from './components/Audio'
export { default as Block } from './components/Block'
export { default as Button } from './components/Button'
export { default as Checkbox } from './components/Checkbox'
export { default as CheckboxGroup } from './components/Checkbox/checkboxGroup'
export { default as Form } from './components/Form'
export { default as Icon } from './components/Icon'
export { default as Image } from './components/Image'
export { default as Input } from './components/Input'
export { default as Label } from './components/Label'
export { default as Map } from './components/Map'
export { default as Picker } from './components/Picker'
export { default as Progress } from './components/Progress'
export { default as Radio } from './components/Radio'
export { default as RadioGroup } from './components/Radio/radioGroup'
export { default as RichText } from './components/RichText'
export { default as ScrollView } from './components/ScrollView'
export { default as Slider } from './components/Slider'
export { default as Swiper } from './components/Swiper'
export { default as SwiperItem } from './components/Swiper/SwiperItem'
export { default as Switch } from './components/Switch'
export { default as Tabbar } from './components/Tabbar'
export { default as TabbarContainer } from './components/Tabbar/Container'
export { default as TabbarPanel } from './components/Tabbar/Panel'
export { default as Text } from './components/Text'
export { default as Textarea } from './components/Textarea'
// export { default as Video } from './components/Video'
export { default as View } from './components/View'
export { default as CoverView } from './components/View'
export { default as CoverImage } from './components/Image'
export { default as WebView } from './components/WebView'
