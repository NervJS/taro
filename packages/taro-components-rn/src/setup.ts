import '@testing-library/jest-native/extend-expect'
