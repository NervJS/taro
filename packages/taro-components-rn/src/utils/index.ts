import {
  StyleSheet
} from 'react-native'

// @see https://facebook.github.io/react-native/docs/layout-props.html
// @see https://facebook.github.io/react-native/docs/view-style-props.html
// @todo According to the source code of ScrollView, ['alignItems','justifyContent'] should be set to contentContainerStyle

const WRAPPER_TYPE_STYLE_REGEX: RegExp = /alignSelf|aspectRatio|border.*|bottom|direction|display|end|left|margin.*|position|right|start|top|zIndex|opacity|elevation/
// const INNER_TYPE_STYLE_REGEX: RegExp = /alignContent|alignItems|flexDirection|flexWrap|height|justifyContent|.*[wW]idth|.*[hH]eight|overflow|padding.*/
const SYNC_TYPE_STYLE_REGEX: RegExp = /flex|flexBasis|flexGrow|flexShrink/

export const omit = (obj: any = {}, fields: string[] = []): { [key: string]: any } => {
  const shallowCopy = { ...obj }
  fields.forEach((key) => {
    delete shallowCopy[key]
  })
  return shallowCopy
}

export const dismemberStyle = (style?: StyleProp<ViewStyle>): { wrapperStyle: ViewStyle; innerStyle: ViewStyle; } => {
  const flattenStyle: ViewStyle & { [key: string]: any } = StyleSheet.flatten(style)
  const wrapperStyle: ViewStyle & { [key: string]: any } = {}
  const innerStyle: ViewStyle & { [key: string]: any } = {}
  if (flattenStyle) {
    Object.keys(flattenStyle).forEach((key: string) => {
      if (SYNC_TYPE_STYLE_REGEX.test(key)) {
        wrapperStyle[key] = flattenStyle[key]
        innerStyle[key] = flattenStyle[key]
      } else if (WRAPPER_TYPE_STYLE_REGEX.test(key)) {
        wrapperStyle[key] = flattenStyle[key]
      } else {
        innerStyle[key] = flattenStyle[key]
      }
    })
  }
  return {
    wrapperStyle,
    innerStyle
  }
}

/**
 * Parses a string of inline styles into a javascript object with casing for react
 *
 * @param {string} styles
 * @returns {Object}
 */
export const parseStyles = (styles: string = ''): { [key: string]: string } => {
  return styles
    .split(';')
    .filter((style) => style.split(':').length === 2)
    .map((style) => [
      style.split(':')[0].trim().replace(/-./g, c => c.substr(1).toUpperCase()),
      style.split(':')[1].trim()
    ])
    .reduce((styleObj, style) => ({
      ...styleObj,
      [style[0]]: style[1],
    }), {})
}

export const noop = (...args: any[]): void => {}

export default {
  omit,
  dismemberStyle,
  parseStyles,
  noop,
}
