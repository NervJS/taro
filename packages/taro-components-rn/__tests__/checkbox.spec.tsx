import * as React from 'react'
import { View, TouchableWithoutFeedback } from 'react-native'
import { shallow } from 'enzyme'
import * as sinon from 'sinon'
import { Label, Checkbox, CheckboxGroup } from '../src'
import * as renderer from 'react-test-renderer'

describe('<Checkbox />', () => {
  describe('<Checkbox />', () => {
    it('default', () => {
      const tree = renderer.create(<Checkbox />).toJSON()
      expect(tree).toMatchSnapshot()
    })

    describe('prop checked changed', () => {
      it('should be checked', () => {
        const wrapper = shallow(
          <Checkbox checked={false} />
        )
        expect(wrapper.state('checked')).toBe(false)
        wrapper.setProps({ checked: true })
        expect(wrapper.state('checked')).toBe(true)
      })
    })

    describe('onPress', () => {
      it('normal', () => {
        const onChange = sinon.spy()
        const wrapper = shallow(
          <Checkbox onChange={onChange} />
        )
        // @ts-ignore
        wrapper.find(TouchableWithoutFeedback).at(0).props().onPress()
        expect(onChange.calledOnce).toBe(true)
        expect(wrapper.state('checked')).toBe(true)
      })

      it('disabled', () => {
        const onChange = sinon.spy()
        const wrapper = shallow(
          <Checkbox onChange={onChange} disabled={true} />
        )
        // @ts-ignore
        wrapper.find(TouchableWithoutFeedback).at(0).props().onPress()
        expect(onChange.calledOnce).toBe(false)
        expect(wrapper.state('checked')).toBe(false)
      })
    })
  })

  describe('<CheckboxGroup />', () => {
    it('should render well', () => {
      const spy = sinon.spy()
      const wrapper = shallow(
        <CheckboxGroup onChange={spy} style={{ flexDirection: 'row' }}>
          <View><View /></View>
          <Label><Checkbox value={0} /></Label>
          <Label><Checkbox value={1} /></Label>
          <Label><Checkbox value={2} /></Label>
        </CheckboxGroup>
      )
      wrapper.find(Checkbox).at(0).props().onChange!({ checked: true, value: 0 })
      expect(spy.calledOnce).toBe(true)
      wrapper.find(Checkbox).at(0).props().onChange!({ checked: false, value: 0 })
      expect(spy.calledTwice).toBe(true)
      wrapper.find(Checkbox).at(1).props().onChange!({ checked: true, value: 1 })
      wrapper.find(Checkbox).at(2).props().onChange!({ checked: true, value: 2 })
      expect(spy.callCount).toBe(4)
    })
  })
})
