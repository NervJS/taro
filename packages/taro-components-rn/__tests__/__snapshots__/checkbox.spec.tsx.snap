// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> <Checkbox /> default 1`] = `
<View
  accessible={true}
  focusable={true}
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      false,
    ]
  }
>
  <View
    hoverStartTime={20}
    hoverStayTime={70}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#f3f3f3",
          "borderRadius": 4,
          "borderWidth": 2,
          "height": 24,
          "justifyContent": "center",
          "width": 24,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      style={
        Array [
          Array [
            Object {
              "opacity": 0,
            },
            false,
          ],
          Object {
            "height": 18,
            "width": 18,
          },
        ]
      }
    >
      <Image
        source={[Function]}
        style={
          Object {
            "height": 18,
            "tintColor": "red",
            "width": 18,
          }
        }
      />
    </View>
  </View>
</View>
`;
